### ISSUE TITLE
Aws::ECS::Client.new.wait_until throws jmespath error when stubbing responses

### ISSUE THREAD

**Issue by apurvis** [aws/aws-sdk-ruby#1307]
tried with both 2.6 and 2.4

## tl;dr recreation steps 
1. clone https://github.com/lumoslabs/broadside
2. `cd broadside && bundle install && bundle exec rspec` 

## longer explanation
I'm trying to make a `wait_until` call on an ECS client object during a spec and getting what looks like a fairly low level exception.

I don't seem to be able to recreate this bug with a simple IRB based approach - when I do this:
```ruby
require 'aws-sdk'
ecs_client = Aws::ECS::Client.new(region: 'us-east-1', credentials: Aws::Credentials.new('access', 'secret'), stub_responses: true)

ecs_client.wait_until(:services_stable, { cluster: 'cluster', services: ['job'] }) do |w|
  w.max_attempts = 60
  w.delay = 2
end
```

i get no error.  but once i stub a few responses and i run [this spec](https://github.com/lumoslabs/broadside/blob/master/spec/broadside/deploy/ecs_deploy_spec.rb#L67-L97) (the one marked `pending`) in the [broadside project](https://github.com/lumoslabs/broadside) from [this branch](https://github.com/lumoslabs/broadside/pull/16) I get this exception which seems extremely low level... such that it should maybe be prevented higher up the call chain?

```
E, [2016-09-27_22:12:17#12859] ERROR -- : #<JMESPath::Errors::InvalidTypeError: function length() expects string, array or object> 
 /Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:42:in `maybe_raise'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:186:in `call'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:23:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/comparator.rb:27:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/and.rb:11:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/not.rb:10:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/condition.rb:11:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/projection.rb:14:in `block in visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/projection.rb:13:in `each'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/projection.rb:13:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:23:in `block in visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:23:in `map'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/function.rb:23:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/nodes/comparator.rb:27:in `visit'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath/runtime.rb:57:in `search'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/jmespath-1.3.1/lib/jmespath.rb:32:in `search'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/poller.rb:70:in `matches_path?'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/poller.rb:65:in `acceptor_matches?'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/poller.rb:49:in `block in call'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/poller.rb:48:in `each'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/poller.rb:48:in `call'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:104:in `block in poll'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:101:in `loop'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:101:in `poll'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:91:in `block (2 levels) in wait'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:90:in `catch'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:90:in `block in wait'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:89:in `catch'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/waiters/waiter.rb:89:in `wait'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/aws-sdk-core-2.6.4/lib/aws-sdk-core/client_waiters.rb:110:in `wait_until'
/Users/stardestroyer/workspace/broadside/lib/broadside/deploy/ecs_deploy.rb:217:in `update_service'
/Users/stardestroyer/workspace/broadside/lib/broadside/deploy/ecs_deploy.rb:29:in `block in deploy'
/Users/stardestroyer/workspace/broadside/lib/broadside/deploy.rb:32:in `deploy'
/Users/stardestroyer/workspace/broadside/lib/broadside/deploy/ecs_deploy.rb:24:in `deploy'
/Users/stardestroyer/workspace/broadside/spec/broadside/deploy/ecs_deploy_spec.rb:98:in `block (5 levels) in <top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/matchers/built_in/raise_error.rb:52:in `call'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/matchers/built_in/raise_error.rb:52:in `matches?'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/matchers/built_in/raise_error.rb:71:in `does_not_match?'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/handler.rb:78:in `does_not_match?'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/handler.rb:72:in `block in handle_matcher'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/handler.rb:70:in `handle_matcher'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/expectation_target.rb:67:in `not_to'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-expectations-3.4.0/lib/rspec/expectations/expectation_target.rb:92:in `not_to'
/Users/stardestroyer/workspace/broadside/spec/broadside/deploy/ecs_deploy_spec.rb:98:in `block (4 levels) in <top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:236:in `instance_exec'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:236:in `block in run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:478:in `block in with_around_and_singleton_context_hooks'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:435:in `block in with_around_example_hooks'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/hooks.rb:478:in `block in run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/hooks.rb:616:in `run_around_example_hooks_for'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/hooks.rb:478:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:435:in `with_around_example_hooks'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:478:in `with_around_and_singleton_context_hooks'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example.rb:233:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:581:in `block in run_examples'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:577:in `map'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:577:in `run_examples'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:543:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `block in run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `map'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `block in run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `map'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/example_group.rb:544:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:119:in `block (3 levels) in run_specs'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:119:in `map'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:119:in `block (2 levels) in run_specs'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/configuration.rb:1680:in `with_suite_hooks'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:118:in `block in run_specs'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/reporter.rb:77:in `report'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:117:in `run_specs'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:93:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:78:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/lib/rspec/core/runner.rb:45:in `invoke'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/rspec-core-3.4.4/exe/rspec:4:in `<top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/bin/rspec:23:in `load'
/Users/stardestroyer/.rbenv/versions/2.1.9/bin/rspec:23:in `<top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:63:in `load'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:63:in `kernel_load'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:24:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/cli.rb:304:in `exec'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/cli.rb:11:in `start'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/exe/bundle:27:in `block in <top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/lib/bundler/friendly_errors.rb:98:in `with_friendly_errors'
/Users/stardestroyer/.rbenv/versions/2.1.9/lib/ruby/gems/2.1.0/gems/bundler-1.12.5/exe/bundle:19:in `<top (required)>'
/Users/stardestroyer/.rbenv/versions/2.1.9/bin/bundle:23:in `load'
/Users/stardestroyer/.rbenv/versions/2.1.9/bin/bundle:23:in `<main>'
```

cc @matl33t

**Comment by awood45** [comment]
It looks like the stack trace (relating to your code) is out of date. Have you been able to isolate a smaller failing example?

I'm looking at this in parallel.

**Comment by apurvis** [comment]
> It looks like the stack trace (relating to your code) is out of date.

it probably is, but you can recreate by checking out the latest and running specs.
```
git clone git@github.com:lumoslabs/broadside.git
git checkout force_task_definition_update
bundle install
bundle exec rspec
```
the spec won't fail because it is `pending` but you will see the stack trace.

> Have you been able to isolate a smaller failing example?

i have not, sorry, though this behavior has been consistent.

**Comment by awood45** [comment]
Okay, I did get a smaller reproducing example:

```ruby
stub = Aws::ECS::Client.new(stub_responses: true)
stub_service_response = {:services=>[{:service_name=>"TEST_TARGET", :service_arn=>"arn:aws:ecs:us-east-1:1234:service/TEST_TARGET"}], :failures=>[]}
stub.stub_responses(:describe_services, stub_service_response)
stub.wait_until(:services_stable, { cluster: 'cluster', services: ['job'] })
```

Will take a look.

**Comment by awood45** [comment]
Okay, I got it. Your stubbed service response lacks the `:deployments` shape in the response. The waiter is raising because it doesn't know how to handle the malformed response. You need to modify your stub like so:

```ruby
stub_service_response = { services: [{ service_name: "TEST_TARGET", service_arn: "arn:aws:ecs:us-east-1:1234:service/TEST_TARGET", deployments: [{desired_count: 1, running_count: 1}] }], failures: [] }
```

That will work. I understand the confusion here, we don't tend to do validation on response shapes since the service doesn't require it - but it does open the possibility of confusion when stubbing. Sorry about that!

**Comment by apurvis** [comment]
Thanks!  I think you do validate some response shapes, but I guess not this one, which is confusing.

While i'm here, is it possible to see the past requests made to the stubbed endpoint in anyway?

Like if a block of code is run, is there a way to see that the calls were like `list_task_instances`, `describe_task_instances(some_arguments)` and maybe see what those arguments were?

**Comment by apurvis** [comment]
thanks that's very helpful
